nengo_loihi/conv.py

```diff
@@ -669,42 +227,59 @@ def conv2d_loihi_weights(conv2d_transform):
 
         weight_key = (tuple(wmask_i), tuple(wmask_j))
         if weight_key not in weights_map:
-            w = kernel[:, wmask_i[:, None]*wmask_j, :]
-            assert w.size == nc * wmask_i.sum() * wmask_j.sum() * nf
-            assert w.shape == (nc, wmask_i.sum() * wmask_j.sum(), nf)
-
-            if output_channels_last:
-                w = w.reshape(nc, -1)
+            # tranpose kernel to (in_channels, rows, cols, out_channels)
+            kernel = np.transpose(transform.init, (2, 0, 1, 3))
+
+            # flip weights to do correlation
+            kernel = kernel[:, ::-1, ::-1, :]
+
+            w = kernel[:, wmask_i[:, None] * wmask_j, :]
+            assert w.size == (inp_shape.n_channels
+                              * wmask_i.sum()
+                              * wmask_j.sum()
+                              * transform.n_filters)
+            assert w.shape == (inp_shape.n_channels,
+                               wmask_i.sum() * wmask_j.sum(),
+                               transform.n_filters)
+
+            if transform.channels_last:
+                w = w.reshape(inp_shape.n_channels, -1)
                 inds = (
-                    np.zeros((nc, 1, 1, 1), dtype=int)
-                    + nyj*nf*np.arange(wmask_i.sum())[:, None, None]
-                    + nf*np.arange(wmask_j.sum())[:, None]
-                    + np.arange(nf)
-                ).reshape(nc, -1)
+                        np.zeros((inp_shape.n_channels, 1, 1, 1), dtype=int)
+                        + (output_cols * transform.n_filters
+                           * np.arange(wmask_i.sum())[:, None, None])
+                        + transform.n_filters * np.arange(wmask_j.sum())[:,
+                                                None]
+                        + np.arange(transform.n_filters)
+                ).reshape(inp_shape.n_channels, -1)
```

Comment:
Personally, I find this harder to read than how it was before. For me, I guess it feels like math; I find math much easier to read when it's all one-letter variable names (possibly with a short subscript, as math is typically written). This new code looks much more intimidating to me, and it's much less clear what's going on. The pattern where each line is being broadcasted to a different row of the matrix gets lost.

Reply:
For me it's the opposite, I find the single-letter variable code very hard to read.  E.g. if I look at `nyf` I have no idea what that is unless I trace its origin all the way back through the code, but `output_cols` gives me a pretty good idea what that variable represents.  We might need a tie-breaker vote on this one 😄 .

Reply:
I've made it clearer to me by making `inds0`, `inds1` etc. variables for each element in the sum, and then adding them together at the end. Take a look and let me know what you think.